### PR TITLE
Allow a user to specify the extraction directory

### PR DESCRIPTION
**Pull request by jacobbogdanov**
Co-authored: @ian-h-chamberlain

TLDR: Add `extract_dir` to `par_binary` which allows reuse of the
unpacked archive results.

When a `par_binary` is launched, if `zip_safe` is set to false, the zip
is unpacked to a temporary directory and deleted on exit.  We have run
into several issues because of this:

- There is an overhead, both CPU and user time, to unpack the zip each
    time that the application is launched.  If the `par_binary` is a
    CLI tool this overhead can be quite noticeable.  Especially if it's
    a tool that is supposed to be small, single function, script which is
    invoked frequently.

- The tempdir isn't cleaned up if the process recieves a KILL signal or
    the app exits uncleanly.  If using `systemd` a possible solution is to
    use `PrivateTmp=true` in the service unit file, which means that the
    entire `/tmp` directory is unique for the service and cleaned up by
    `systemd` when the application terminates (in any possible way.)

- No way to modify the archive on the fly.  This one is a questionable
    limitation as rebuilding and re-deploying the archive is a quick
    alternative.  We have explored whether a solution similar to
    https://superuser.com/questions/647674/is-there-a-way-to-edit-files-inside-of-a-zip-file-without-explicitly-extracting
    are possible.

The solution: Extract the archive to a static location, and re-use the
artifacts if the `par_binary` has not changed.

- Add a new `extract_dir` argument to the `par_binary` (and `par_test`,)
    macros.  This is a file path to where the the par archive will be
    unpacked to.  `extract_dir` is only applicable if `zip_safe == False`.

- **The `extract_dir` is not cleaned up on exit.**  If the archive has
    not changed then subsequent invocations will reuse the same unpacked
    results.  This saves the computational load of extracting the par each
    time that it's invoked.

- If the `extract_dir` cannot be written to then the application will
    fall back to writing to a temporary directory.

- If the `par_binary` changes the `extract_dir` will be removed and
    recreated.  **If the files within the `extract_dir` are modified the
    par won't be recreated.**

- Add a new runtime environment variable `UNPAR_EXTRACT_ONLY`. It
    launches the par, unpacks it to the `extract_dir`, and exits
    immediately.

**WARNING** Setting an `extract_dir` makes the `par_binary` unsafe to
invoke from multiple threads as the first invocation needs to unpack
into the `extract_dir`. To get around this invoke the application with
`UNPAR_EXTRACT_ONLY` during installation.

### CODE REVIEW

**Review by ian-h-chamberlain — APPROVED** [review]

File: test_dir_shadowing/test_dir_shadowing_main_PY2_filelist.txt [hunk 1]
```diff
@@ -9,3 +9,4 @@ subpar/test_dir_shadowing/test_dir_shadowing/dir_shadowing_lib_dat.txt
 subpar/test_dir_shadowing/test_dir_shadowing/dir_shadowing_main.py
 subpar/test_dir_shadowing/test_dir_shadowing/dir_shadowing_main_dat.txt
 subpar/test_dir_shadowing/test_dir_shadowing_main
+UNPAR_MANIFEST
```
Comment (jacobbogdanov) [inline test_dir_shadowing/test_dir_shadowing_main_PY2_filelist.txt:12]: do you think we should only add the `UNPAR_MANIFEST` if `extract_dir` is set?

File: runtime/support.py [hunk 2]
```diff
@@ -309,7 +365,25 @@ def setup(import_roots, zip_safe):
 
     # Extract files to disk if necessary
     if not zip_safe:
-        extract_dir = _extract_files(archive_path)
+        extract_only = os.environ.get("UNPAR_EXTRACT_ONLY")
```
Comment (jacobbogdanov) [inline runtime/support.py:368]: Maybe we should drop the `UN` prefix... I know we've been calling this feature unpar, but both in the manifest and in the environment variables it might make sense to just have it be `PAR_` prefix

File: test_dir_shadowing/test_dir_shadowing_main_PY2_filelist.txt (hunk 1 above)
Comment (ian-h-chamberlain) [reply test_dir_shadowing/test_dir_shadowing_main_PY2_filelist.txt:12]: Good point, it's pretty much completely extraneous in the non-unpar case. I can make that change along with the next PR

File: runtime/support.py (hunk 2 above)
Comment (ian-h-chamberlain) [reply runtime/support.py:368]: Yeah, particularly for the environment variable it doesn't make as much sense. The manifest file I think can be whatever we want, but I suppose you're right in that it's really a manifest of what's in the par, not what's in the unpar'd directory.
